### PR TITLE
REFACTOR: helpers for take_2d_labels_or_positional

### LINKED ISSUE #4922: REFACTOR: `take_2d_labels_or_positional` code refactoring
Place: https://github.com/modin-project/modin/blob/84f638cc703beb550edecce715532c8e86fb7b25/modin/core/dataframe/pandas/dataframe/dataframe.py#L617

### PR DESCRIPTION
**Pull request by jbrockmendel**
## What do these changes do?

- [ ] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/development/contributing.html#commit-message-formatting)
- [ ] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [ ] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [ ] signed commit with `git commit -s` 
- [x] Resolves #4922 
- [ ] tests added and passing
- [ ] module layout described at `docs/development/architecture.rst` is up-to-date 
- [ ] added (Issue Number: PR title (PR Number)) and github username to release notes for next major release

### CODE REVIEW

**Comment by jbrockmendel** [comment]
Looks like `from __future__ import annotations` isnt supported in py36.  any plans to drop support for that?  pandas is already py38+

File: docs/release_notes/release_notes-0.16.0.rst [hunk 1]
```diff
@@ -83,6 +83,7 @@ Key Features and Updates
   * REFACTOR-#3780: Remove code duplication for `PandasOnDaskDataframe` (#3781)
   * REFACTOR-#4530: Unify access to physical data for any partition type (#4829)
   * REFACTOR-#4885: De-duplicated take_2d_labels_or_positional methods (#4883)
+  * REFACTOR-#0000: Helpers for take_2d_labels_or_positional (#4865)
```
Comment (anmyachev) [inline docs/release_notes/release_notes-0.16.0.rst:86]:
```suggestion
  * REFACTOR-#4922: Helpers for take_2d_labels_or_positional (#4865)
```

File: modin/core/dataframe/pandas/dataframe/dataframe.py [hunk 2]
```diff
@@ -647,12 +649,132 @@ def take_2d_labels_or_positional(
...
+        Sequence[int]
+        """
+        # Helper for take_2d_positional
+        if is_range_like(positions) and positions.step > 0:
+            sorted_positions = positions
+        else:
+            sorted_positions = np.sort(positions)
+        return sorted_positions
+
+    def _get_new_lengths(self, partitions_dict, *, axis: int) -> List[int]:
```
Comment (anmyachev) [inline modin/core/dataframe/pandas/dataframe/dataframe.py:690]: Why is an asterisk used here?

File: modin/core/dataframe/pandas/dataframe/dataframe.py [hunk 3]
```diff
@@ -647,12 +649,132 @@ def take_2d_labels_or_positional(
...
+        # Use the slice to calculate the new columns
+        if axis == 0:
+            idx = self.index
+        else:
+            idx = self.columns
+
+        # TODO: Support fast processing of negative-step ranges
+        if is_range_like(positions) and positions.step > 0:
+            # pandas Index is more likely to preserve its metadata if the indexer
+            #  is slice
```
Comment (anmyachev) [inline modin/core/dataframe/pandas/dataframe/dataframe.py:749]:
```suggestion
            # is slice
```

File: modin/core/dataframe/pandas/dataframe/dataframe.py [hunk 4]
```diff
@@ -647,12 +649,132 @@ def take_2d_labels_or_positional(
...
+            monotonic_idx = slice(positions.start, positions.stop, positions.step)
+        else:
+            monotonic_idx = np.asarray(sorted_positions, dtype=np.intp)
+
+        new_idx = idx[monotonic_idx]
+        return new_idx, monotonic_idx
+
+    @lazy_metadata_decorator(apply_axis=None)
```
Comment (anmyachev) [inline modin/core/dataframe/pandas/dataframe/dataframe.py:757]:
```suggestion
```

**Review by anmyachev — CHANGES_REQUESTED** [review]

File: modin/core/dataframe/pandas/dataframe/dataframe.py (hunk 2 above)
Comment (jbrockmendel) [reply modin/core/dataframe/pandas/dataframe/dataframe.py:690]: makes axis keyword-only

Comment (anmyachev) [reply modin/core/dataframe/pandas/dataframe/dataframe.py:690]: Got it

**Review by anmyachev — APPROVED** [review]
LGTM!

**Pull request merged**